### PR TITLE
USHIFT-6913: Add DNS resource configuration enhancement for MicroShift

### PR DESCRIPTION
**Pull request by Neilhamza**
## Summary
  - Adds enhancement proposal for [OCPSTRAT-3015](https://redhat.atlassian.net/browse/OCPSTRAT-3015): MicroShift DNS deployment resource configuration                                                                                 
  - Introduces `dns.resources` config option to override CPU/memory requests and limits for the `dns` container in the `dns-default` DaemonSet
  - Follows the same pattern as the existing `dns.hosts` feature

### CODE REVIEW

File: enhancements/microshift/microshift-dns-resource-configuration.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+```go
+type DNS struct {
+    BaseDomain string        `json:"baseDomain"`
+    Hosts      HostsConfig   `json:"hosts,omitempty"`
+    Resources  *DNSResources `json:"resources,omitempty"`
+}
+
+type DNSResources struct {
+    Requests map[string]string `json:"requests,omitempty"`
+    Limits   map[string]string `json:"limits,omitempty"`
```
Comment (pacevedom) [inline enhancements/microshift/microshift-dns-resource-configuration.md:133]: Should we use specific fields here instead of arbitrary keys? With current types a user could input `gpus: 1` for example.
Wondering if `corev1.ResourceRequirements` type would bring any benefits too.

File: enhancements/microshift/microshift-dns-resource-configuration.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  resources:
+    requests:
+      cpu: <resource.Quantity>    # optional, default: 50m
+      memory: <resource.Quantity> # optional, default: 70Mi
+    limits:
+      cpu: <resource.Quantity>    # optional, default: not set
+      memory: <resource.Quantity> # optional, default: not set
+```
+
+By default, the `dns.resources` section is not set, and the current hardcoded defaults are used. When provided, individual fields within `requests` and `limits` are optional - users can set only the values they want to override.
```
Comment (pacevedom) [inline enhancements/microshift/microshift-dns-resource-configuration.md]: If a user includes any of those values, does that invalidate the default setting for the others? For example, setting cpu to 500m still keeps the 70Mi for memory, or does it become unset?

File: enhancements/microshift/microshift-dns-resource-configuration.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+Extend `startDNSController()` in `pkg/components/controllers.go` to pass resource values as extra render parameters, with defaults matching the current hardcoded values:
+
+```go
+extraParams := assets.RenderParams{
+    "ClusterIP":        cfg.Network.DNS,
+    "HostsEnabled":     cfg.DNS.Hosts.Status == config.HostsStatusEnabled,
+    "DNSCPURequest":    dnsResourceValue(cfg.DNS.Resources, "requests", "cpu", "50m"),
+    "DNSMemoryRequest": dnsResourceValue(cfg.DNS.Resources, "requests", "memory", "70Mi"),
+    "DNSCPULimit":      dnsResourceValue(cfg.DNS.Resources, "limits", "cpu", ""),
+    "DNSMemoryLimit":   dnsResourceValue(cfg.DNS.Resources, "limits", "memory", ""),
```
Comment (pacevedom) [inline enhancements/microshift/microshift-dns-resource-configuration.md]: Unset limits would render empty strings, which renders to false in the template. This is now a requirement for the function, right?

File: enhancements/microshift/microshift-dns-resource-configuration.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+}
+```
+
+### Risks and Mitigations
+
+**Risk:** Users configure resources too low, causing CoreDNS to be OOM-killed or throttled.
+**Mitigation:** Document minimum recommended values. MicroShift does not enforce minimum thresholds to allow flexibility on constrained devices.
```
Comment (pacevedom) [inline enhancements/microshift/microshift-dns-resource-configuration.md]: We might need to include a minimum setting or else the cluster is not even able to start properly (without dns there is no readiness from all pods).

File: enhancements/microshift/microshift-dns-resource-configuration.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+**Risk:** Users configure resource limits without requests, leading to unexpected Kubernetes scheduling behavior.
```
Comment (lance5890) [inline enhancements/microshift/microshift-dns-resource-configuration.md]: I'm afraid that if u only set the limit value without setting the request value, the request value should be equal to the limit value by default

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 1 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md:133]: restrict keys to only "cpu" and "memory" during validation, unsupported keys like "gpu" are rejected with a clear error.  in the implementation we used map[string]string to stay consistent with the existing config patterns (simple Go types, not Kubernetes API types). Using corev1.ResourceRequirements would be the only Kubernetes type in the config package and would add coupling to the API types for serialization.

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 2 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: setting only cpu: 500m preserves the default memory: 70Mi. The implementation uses key-by-key merge, not full map replacement. and we have a test for it

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 3 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Yes, unset limits return empty strings. 
The template uses {{- if .DNSHasLimits }} (checks len(cfg.DNS.Resources.Limits) > 0) to conditionally render the entire limits block. Individual limit values are also guarded with {{- if .DNSCPULimit }}. So empty strings never appear in the rendered YAML.

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 4 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Good point. We could add minimum thresholds in validation (e.g., reject cpu < 10m or memory < 20Mi), but  could this block legitimate use cases on extremely constrained edge devices?

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 5 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: In the implementation, default requests (cpu=50m, memory=70Mi) are always populated by dnsDefaults(). If a user sets only limits, the defaults remain as  requests. The rendered DaemonSet YAML always contains explicit requests and limits values
adding this to the Document

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 1 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md:133]: Added to the Document that only cpu and memory keys are accepted, unsupported keys rejected during validation

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 2 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Added to the Document explicit explanation of key-by-key merge behavior and that defaults are preserved

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 3 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Updated template to use DNSHasLimits boolean (matching implementation), added explanation of nil map         behavior

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 4 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: updated mitigation text, acknowledged the risk, noted minimum validation can be added later

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 5 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Added to the Document detailed explanation that default requests are always populated, and validation catches limit default request

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 4 above)
Comment (pacevedom) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: We have minimum requirements to run MicroShift (without including the app). Even if there is an extreme resource contraint the minimum requirements must hold, and having too low resources for dns might render unstable clusters (dns not responding, throttling in other pods, etc.).
What do you think about having the minimum the same as the default value we ship, as that has been the case so far since first version?

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 5 above)
Comment (pacevedom) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Should we add a test for this in the test plan? This is a corner case for the config.

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 4 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: sure! updated the minimum requirements same as the default we ship

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 5 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: yes! added the test in the e2e tests and in the test plan

File: enhancements/microshift/microshift-dns-resource-configuration.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+Edge deployments have diverse resource profiles. Devices running high DNS query volumes or custom CoreDNS plugins may need more CPU and memory than the defaults.
+Conversely, extremely constrained devices may benefit from reducing the resource requests to free capacity for application workloads.
+Currently, the only way to change DNS pod resources is to manually patch the DaemonSet after MicroShift starts, which is overwritten on the next restart.
+Customers need a supported, persistent way to tune DNS pod resources through the MicroShift configuration file.
+
+### User Stories
+
+1. As a MicroShift administrator running high DNS query workloads, I want to increase the CPU and memory allocated to CoreDNS so that DNS resolution remains responsive under load.
+2. As a MicroShift administrator on a resource-constrained edge device, I want to reduce the DNS pod resource requests so that more capacity is available for application workloads.
```
Comment (pacevedom) [inline enhancements/microshift/microshift-dns-resource-configuration.md]: Should we reword this? I would drop the `reduce` to mention something about adjusting DNS resources so that there are no disruptions to the service. This is now at odds with non-goal [#4](https://github.com/openshift/enhancements/pull/1989/changes#diff-ab9c896b5d0d5d7e64ad716a4fc08f9f92da4d6387e468a905c6dc9de3ee72c7R55)

File: enhancements/microshift/microshift-dns-resource-configuration.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+When upgrading from a version without `dns.resources` support, the new configuration field will remain unset, causing the existing defaults to be used. No user action is required on upgrade.
+
+When downgrading to a version without `dns.resources` support, the field is ignored by the older version. MicroShift reverts to the hardcoded default resources on startup. The configuration file on disk is not modified.
+
+## Version Skew Strategy
+N/A
+
+## Operational Aspects of API Extensions
+N/A
```
Comment (pacevedom) [inline enhancements/microshift/microshift-dns-resource-configuration.md]: Maybe mention something about setting limits and getting OOM kills? This is k8s default behavior though, so up to you.

**Review by pacevedom — COMMENTED** [review]
2 nits

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 7 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Done

File: enhancements/microshift/microshift-dns-resource-configuration.md (hunk 6 above)
Comment (Neilhamza) [reply enhancements/microshift/microshift-dns-resource-configuration.md]: Done

**Review by pacevedom — APPROVED** [review]
/lgtm

Thank you!
